spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala

```diff
@@ -0,0 +1,143 @@
+
+  private val FILE_NAME_COL = "_file"
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    // no need to execute this delete as the condition is false
+    case DeleteFromTable(_, Some(FalseLiteral)) =>
+      LocalRelation()
+
+    // don't rewrite deletes that can be answered by passing filters to deleteWhere in SupportsDelete
+    case d @ DeleteFromTable(r: DataSourceV2Relation, Some(cond)) if isDeleteWhereCase(r, cond) =>
+      d
+
+    // rewrite all operations that require reading the table to delete records
+    case DeleteFromTable(r: DataSourceV2Relation, Some(cond)) =>
+      // TODO: do a switch based on whether we get BatchWrite or DeltaBatchWrite
+      val writeInfo = newWriteInfo(r.schema)
+      val mergeBuilder = r.table.asMergeable.newMergeBuilder(writeInfo)
+
+      val (scan, scanPlan) = buildScanPlan(r.table, r.output, mergeBuilder, cond)
+
+      val remainingRowFilter = Not(EqualNullSafe(cond, Literal(true, BooleanType)))
+      val remainingRowsPlan = Filter(remainingRowFilter, scanPlan)
+
+      val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+      batchWrite match {
+        case w: RequiresScan => w.withScan(scan)
+        case _ => // do nothing
+      }
+
+      // TODO: group by something so that we can write back
+      // Option 1: repartition/sort by partition values
+      // Option 2: repartition by _file and local sort by partition values (how to deal with 1 GB files?)
+      ReplaceData(r, batchWrite, remainingRowsPlan)
+  }
+
+  private def buildScanPlan(
+      table: Table,
+      output: Seq[AttributeReference],
+      mergeBuilder: MergeBuilder,
+      cond: Expression): (Scan, LogicalPlan) = {
+
+    val scanBuilder = mergeBuilder.asScanBuilder
+
+    val predicates = splitConjunctivePredicates(cond)
+    val normalizedPredicates = DataSourceStrategy.normalizeExprs(predicates, output)
+    PushDownUtils.pushFilters(scanBuilder, normalizedPredicates)
+
+    val scan = scanBuilder.build()
+    val scanRelation = DataSourceV2ScanRelation(table, scan, output)
+
+    scan match {
+      case _: SupportsFileFilter =>
+        val matchingFilePlan = buildFileFilterPlan(cond, scanRelation)
+        val dynamicFileFilter = DynamicFileFilter(scanRelation, matchingFilePlan)
+        (scan, dynamicFileFilter)
+      case _ =>
+        (scan, scanRelation)
+    }
+  }
+
+  private def isDeleteWhereCase(relation: DataSourceV2Relation, cond: Expression): Boolean = {
+    relation.table match {
+      case t: ExtendedSupportsDelete if !SubqueryExpression.hasSubquery(cond) =>
+        val predicates = splitConjunctivePredicates(cond)
+        val normalizedPredicates = DataSourceStrategy.normalizeExprs(predicates, relation.output)
+        val dataSourceFilters = toDataSourceFilters(normalizedPredicates)
+        val allPredicatesTranslated = normalizedPredicates.size == dataSourceFilters.length
+        allPredicatesTranslated && t.canDeleteWhere(dataSourceFilters)
```

Comment:
We should open a PR against Spark for this. It may be easy to get it into 3.1.0.

Reply:
Definetely.